### ISSUE TITLE
As a software developer of scientific software, I need a native Haskell implementation of the dscal function so that I can multiply a vector by a scalar in a type-safe, thread safe manner.

### ISSUE THREAD

**Issue by dlewissandy** [dlewissandy/lambda-blas#70]
Documentation for the dscal function can be found at BLAS.

- [x] create native implementation of function
- [x] create haddock documentation of function
- [x] create unit test to demonstrate byte equivalence to FORTRAN BLAS library function
- [x] create benchmark(s) to quantify time performance relative to FORTRAN BLAS library function
- [x] demonstrate that native has a relative time performance of no more than 100x the FORTRAN blas function.

**Comment by dlewissandy** [comment]
Implementation Strategy:
* This function can be implemented easily by polymorphising the single precision version.

**Comment by dlewissandy** [comment]
Attached is the raw benchmark data:
[dscal.zip](https://github.com/dlewissandy/lambda-blas/files/868039/dscal.zip)

**Comment by dlewissandy** [comment]
# DSCAL BENCHMARK ANALYSIS #

## Special Case inc=1 ##
When the increment of traversal is equal to unity, the time performance of both the unsafe and pure implementations are well described by a linear model:

### PURE ###
```
lm(formula = PURE ~ N, data = dat[dat$INC == 1, ])

Residuals:
    Min      1Q  Median      3Q     Max
-0.7922 -0.1159  0.1321  0.1416  1.2848

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -9.402e-02  6.039e-02  -1.557    0.128
N            2.348e-03  9.180e-06 255.797   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3407 on 37 degrees of freedom
Multiple R-squared:  0.9994,	Adjusted R-squared:  0.9994
F-statistic: 6.543e+04 on 1 and 37 DF,  p-value: < 2.2e-16
```
### UNSAFE ###
```
lm(formula = UNSAFE ~ N, data = dat[dat$INC == 1, ])

Residuals:
    Min      1Q  Median      3Q     Max
-0.6309 -0.1033  0.1207  0.1524  0.6541

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -7.315e-02  4.465e-02  -1.638     0.11
N            9.253e-04  6.787e-06 136.334   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.2519 on 37 degrees of freedom
Multiple R-squared:  0.998,	Adjusted R-squared:  0.998
F-statistic: 1.859e+04 on 1 and 37 DF,  p-value: < 2.2e-16
```
### ASYMPTOTIC PERFORMANCE ###
In the limit as n becomes very large, the relative time performance should approach:
```
2.348e-03
-------- = 2.5x
9.253e-04
```
which is within the 100x performance threshold for this project milestone

## Special Case inc/=1 ##
When the increment of traversal is not equal to unity, the time performance of both the unsafe and pure implementations are well described by a multi-linear model:

### PURE ###
```
lm(formula = PURE ~ N + Z, data = dat[dat$INC != 1, ])

Residuals:
     Min       1Q   Median       3Q      Max
-0.76297 -0.14254  0.05225  0.17111  0.66877

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -1.350e-01  3.985e-02  -3.389  0.00141 **
N            5.554e-03  9.297e-05  59.742  < 2e-16 ***
Z            7.388e-04  7.105e-06 103.986  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.2422 on 48 degrees of freedom
Multiple R-squared:  0.999,	Adjusted R-squared:  0.999
F-statistic: 2.44e+04 on 2 and 48 DF,  p-value: < 2.2e-16
```
### UNSAFE ###
```
lm(formula = UNSAFE ~ N + Z, data = dat[dat$INC != 1, ])

Residuals:
     Min       1Q   Median       3Q      Max
-0.49854 -0.15185  0.05078  0.14867  0.65518

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -9.321e-02  3.803e-02  -2.451   0.0179 *
N            3.707e-03  8.873e-05  41.772   <2e-16 ***
Z            4.404e-04  6.781e-06  64.945   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.2312 on 48 degrees of freedom
Multiple R-squared:  0.9977,	Adjusted R-squared:  0.9976
F-statistic: 1.032e+04 on 2 and 48 DF,  p-value: < 2.2e-16
```
### ASYMPTOTIC PERFORMANCE ###
In the limit as both n and inc become very large, the relative time performance should approach:
```
7.388e-04
--------- = 1.7x
4.404e-04
```
which is within the 100x performance threshold for this project milestone.

**Comment by dlewissandy** [comment]
Closed as DONE after approval of pull request #80.
